### ISSUE TITLE
Fix CI configuration and install proper dependencies on test machines

### ISSUE THREAD

**Issue by pelesh** [LLNL/hiop#107]
It seems that CI fails because HiOp has added new dependencies, which are not installed on test platforms. We need to install proper dependencies and verify that CI runs correctly.

@cnpetra @nychiang

**Comment by nychiang** [comment]
@pelesh Could you show me which dependencies fail CI? The current cmake script checks METIS and COINHSL, but it should work if these two are not found on the platform.

**Comment by ashermancinelli** [comment]
This is the error message for current develop branch (commit a308a478b678fa81dbe48e75f0ae5dfbb7d80705) on ornl ci:

```
-- Found metis include: /autofs/nccsopen-svm1_sw/ascent/.swci/1-compute/opt/spack/20180914/linux-rhel7-ppc64le/gcc-7.4.0/metis-5.1.0-a5sdngk67b5bpxgvz5ki7f4yauvmc3ks/include
-- Found metis library: /autofs/nccsopen-svm1_sw/ascent/.swci/1-compute/opt/spack/20180914/linux-rhel7-ppc64le/gcc-7.4.0/metis-5.1.0-a5sdngk67b5bpxgvz5ki7f4yauvmc3ks/lib/libmetis.so
CMake Error at cmake/FindHiopMETIS.cmake:36 (add_library):
  add_library cannot create target "METIS" because another target with the
  same name already exists.  The existing target is an interface library
  created in source directory "/gpfs/wolf/proj-shared/csc359/ci/125970".  See
  documentation for policy CMP0002 for more details.
Call Stack (most recent call first):
  CMakeLists.txt:203 (include)
-- Found METIS library: /autofs/nccsopen-svm1_sw/ascent/.swci/1-compute/opt/spack/20180914/linux-rhel7-ppc64le/gcc-7.4.0/metis-5.1.0-a5sdngk67b5bpxgvz5ki7f4yauvmc3ks/lib/libmetis.so
-- COINHSL was not found.
-- Cannot find COINHSL nor STRUMPACK for sparse linear algebra, and the option HIOP_SPARSE will be disabled
-- Configuring incomplete, errors occurred!
```

**Comment by ashermancinelli** [comment]
I believe the issue is due to the line `include(FindHiopMETIS)` occuring twice in the top level CMakeLists.txt.

**Comment by nychiang** [comment]
@ashermancinelli yes, that cause the problem. I have fixed it. You can see my email for a quick way to fix it.

**Comment by ashermancinelli** [comment]
I am under the impression that METIS is a mandatory dependency, in which case some of the new CMake logic is a bit erroneous - @pelesh is this the case? If so, I can clean up cmake and fix this.

**Comment by pelesh** [comment]
METIS used to be a "second-hand" dependency. Only SuiteSparse/UMFPACK depended on it.

@nychiang, where do you need METIS?

**Comment by nychiang** [comment]
No, it is not a mandatory dependency. If umfpack and other sparse linear solvers are not used, we don't need METIS.
For example, if we set HIOP_WITH_KRON_REDUCTION = OFF and HIOP_SPARSE = OFF, METIS is not required anymore.
The problem is umfpack (HIOP_WITH_KRON_REDUCTION =ON), coinhsl (HIOP_SPARSE=ON & HIOP_USE_COINHSL  =ON) or strumpack(HIOP_SPARSE=ON & HIOP_USE_STRUMPACK  =ON) need to use METIS.

**Comment by pelesh** [comment]
@nychiang could you please specify what libraries depend on METIS (regardless if these are optional or required HiOp dependencies)?

**Comment by nychiang** [comment]
So far there are 3 libraries:
libumfpack
libcoinhsl
libstrumpack

**Comment by pelesh** [comment]
@ashermancinelli, I think the fix would be to check if METIS was found (and if not, find METIS) for each of the three packages depending on it. Checking if METIS was found first would prevent creating multiple METIS targets.

**Comment by ashermancinelli** [comment]
I have pushed some preliminary fixes to branch `ci-cmake-fix#107` which correct this, however it seems there are some deeper issues here. I will am tracking these issue in #109 as mentioned above.

**Comment by nychiang** [comment]
I have pushed some preliminary fixes last night in d81181e in branch dev-fix#79.
I also correct some other bugs in the later commits for branch dev-fix#79, including the missing lapack module for essl. 
Now we don't need to provide -DLAPACK_LIBRARIES when essl is the default option for lapack.

**Comment by nychiang** [comment]
There is one more issue here as I can build all the hiop examples, e.g., EX1~EX7, but failed to build the tests under the subfolder tests. Will open as a new issue.

**Comment by pelesh** [comment]
> There is one more issue here as I can build all the hiop examples, e.g., EX1~EX7, but failed to build the tests under the subfolder tests. Will open as a new issue.

You need to resync with `develop`.  This was fixed there.

**Comment by nychiang** [comment]
I will add my other fixes for camke into branch ci-cmake-fix#107 and push it to repos.
Then I will rebase my dev-fix#79 to ci-cmake-fix#107, to focus on issue #79.

**Comment by nychiang** [comment]
> > There is one more issue here as I can build all the hiop examples, e.g., EX1~EX7, but failed to build the tests under the subfolder tests. Will open as a new issue.
> 
> You need to resync with `develop`. This was fixed there.

Nope. I tried to rebase on to the latest `develop` or `ci-cmake-fix#107`, and neither solves the problem.

**Comment by pelesh** [comment]
> > > There is one more issue here as I can build all the hiop examples, e.g., EX1~EX7, but failed to build the tests under the subfolder tests. Will open as a new issue.
> > 
> > 
> > You need to resync with `develop`. This was fixed there.
> 
> Nope. I tried to rebase on to the latest `develop` or `ci-cmake-fix#107`, and neither solves the problem.

Please create another issue, so we can figure out what is going on in here.

**Comment by ashermancinelli** [comment]
> I will add my other fixes for camke into branch ci-cmake-fix#107 and push it to repos.
> Then I will rebase my dev-fix#79 to ci-cmake-fix#107, to focus on issue #79.

I don't think you'll need to rebase - what does your diff provide that is not already in `ci-cmake-fix#107`?

**Comment by nychiang** [comment]
I used different way to fix the multiple METIS problem, and hence will introduce some conflicts with your `ci-cmake-fix#107`.
Hence rebase to yours can let me fix this issue now and get the latest 'develop' features.
Based on your fix to the cmake files, there is one more place in cmake files that needed to be fixed.
in `FindHiopCudaLibraries`.cmake, change `FindMagma.cmake` to `FindHiopMagma.cmake`.

**Comment by ashermancinelli** [comment]
@nychiang Does this explicitly have to do with branch `sparse-develop`? If so, I'll give it the appropriate label.

**Comment by nychiang** [comment]
I think you have already fixed some issue in `ci-cmake-fix#107`
Could you test the latest `sparse-develop` again on your platform. Due to #119, I cannot see the results from PNNL pipeline.
I tried to build Hiop with HIOP_USE_RAJA, HIOP_USE_UMPIRE and HIOP_WITH_KRON_REDUCTION on Summit and it looks fine.
However, in order to compile the code, I have to fix something quickly in [matrixTestsRajaSparseTriplet.hpp](https://github.com/LLNL/hiop/commit/e929e63590619d9101b4c2a960ba5ac55f157019#diff-ec06db2f439eb48e64a119ea9e92d38aba155b7c6e128ce9615f32383d7ee089).

**Comment by pelesh** [comment]
We have CI configured on ascent, marianas and newell testing build with the sparse solver enabled and running test examples 6 and 7. @ashermancinelli and @nychiang, I suggest we close this issue and create a new one to install and enable STRUMPACK tests.
